compiler/qsc_circuit/src/circuit_to_qsharp.rs

```diff
@@ -0,0 +1,379 @@
+
+fn generate_qubit_validation(
+    circuit_name: &str,
+    qubits: &FxHashMap<usize, String>,
+    indentation_level: usize,
+) -> String {
+    if qubits.is_empty() {
+        return String::new();
+    }
+
+    let indent = "    ".repeat(indentation_level);
+    let inner_indent = "    ".repeat(indentation_level + 1);
+    format!(
+        "{indent}if Length(qs) != {} {{\n\
+        {inner_indent}fail \"Invalid number of qubits. Operation {} expects a qubit register of size {}.\";\n\
+        {indent}}}\n",
+        qubits.len(),
+        circuit_name,
+        qubits.len()
+    )
+}
+
+fn handle_measurement(
+    measurement: &Measurement,
+    qubits: &FxHashMap<usize, String>,
+    indent: &str,
+    measure_results: &mut Vec<(String, (usize, usize))>,
+) -> String {
+    let operation_str = measurement_call(measurement, qubits);
+    let mut op_results = vec![];
+    for reg in &measurement.results {
+        if let Some(c_id) = reg.result {
+            let result = (format!("c{}_{}", reg.qubit, c_id), (reg.qubit, c_id));
+            op_results.push(result.clone());
+        }
+    }
+
+    // Sort first by q_id, then by c_id
+    op_results.sort_by_key(|(_, (q_id, c_id))| (*q_id, *c_id));
+    let result = op_results
+        .iter()
+        .map(|(name, _)| name.as_str())
+        .collect::<Vec<_>>()
+        .join(", ");
+    match op_results.len() {
+        0 => {
+            format!("{indent}{operation_str};\n")
+        }
+        1 => {
+            measure_results.extend(op_results);
+            format!("{indent}let {result} = {operation_str};\n")
+        }
+        _ => {
+            measure_results.extend(op_results);
+            format!("{indent}let ({result}) = {operation_str};\n")
+        }
+    }
+}
+
+fn handle_unitary(unitary: &Unitary, qubits: &FxHashMap<usize, String>, indent: &str) -> String {
+    // "SX" will generate two operations: Rx and Adj R
+    if unitary.gate == "SX" {
+        let pi_arg = "π / 2.0".to_string();
+        let rx_str = operation_call(
+            &Unitary {
+                gate: "Rx".to_string(),
+                args: vec![pi_arg.clone()],
+                children: vec![],
+                targets: unitary.targets.clone(),
+                controls: unitary.controls.clone(),
+                is_adjoint: unitary.is_adjoint,
+            },
+            qubits,
+        );
+        let general_r_str = operation_call(
+            &Unitary {
+                gate: "R".to_string(),
+                args: vec!["PauliI".to_string(), pi_arg],
+                children: vec![],
+                targets: unitary.targets.clone(),
+                controls: unitary.controls.clone(),
+                is_adjoint: !unitary.is_adjoint, // Adj R
+            },
+            qubits,
+        );
+        if unitary.is_adjoint {
+            format!("{indent}{general_r_str};\n{indent}{rx_str};\n")
+        } else {
+            format!("{indent}{rx_str};\n{indent}{general_r_str};\n")
+        }
```

Comment:
I had to think about this one quite a bit... technically it is correct to reverse the operation order when doing the adjoint, but in this case the phase applied by `R(PauliI, _, _)` and the rotation performed by `Rx` commute with each other. So you could simplify this code to just always emit the `Rx` followed by `R` instead of changing the order based on `unitary.is_adjoint`.

Reply:
I've changed this to the more straightforward decomp we talked about with HSH, which also does not need reversing.